great_expectations/expectations/metrics/util.py

```diff
@@ -1058,3 +1059,83 @@ def get_sqlalchemy_source_table_and_schema(
         )
     else:
         return engine.batch_manager.active_batch_data.selectable
+
+
+def get_unexpected_indices_for_multiple_pandas_named_indices(
+    df: pd.DataFrame,
+    expectation_domain_column_name: str,
+    unexpected_index_column_names: List[str],
+) -> List[Dict[str, Any]]:
+    """
+    Builds unexpected_index list for Pandas Dataframe in situation where the named
+    columns is also a named index. This method handles the case when there are multiple named indices.
+    Args:
+        df: reference to Pandas dataframe
+        expectation_domain_column_name: column that Expectation is being run for
+        unexpected_index_column_names: column_names for indices, either named index or unexpected_index_columns
+
+    Returns:
+        List of Dicts that contain ID/PK values
+    """
+    if not expectation_domain_column_name:
+        raise ge_exceptions.MetricResolutionError(
+            message=f"Error: The domain column is currently set to None. Please check your configuration.",
+            failed_metrics=["unexpected_index_list"],
+        )
+    index_names: List[str] = df.index.names
+    unexpected_index_list: List[Dict[str, Any]] = list()
+    unexpected_indices: List[Union[int, str]] = list(df.index)
+    for index in unexpected_indices:
+        primary_key_dict: Dict[str, Any] = dict()
+        # domain column first
+        primary_key_dict[expectation_domain_column_name] = df.at[
+            index, expectation_domain_column_name
+        ]
+        for column_name in unexpected_index_column_names:
+            if column_name not in index_names:
+                raise ge_exceptions.MetricResolutionError(
+                    message=f"Error: The column {column_name} does not exist in the named indices. Please check your configuration",
+                    failed_metrics=["unexpected_index_list"],
+                )
+            tuple_index = index_names.index(column_name, 0)
```

Comment:
This logic is identical for each loop through `unexpected_indices` so I'd do this validation once at the beginning of this function outside the `unexpected_indices` loop since that loop may be large.

That is, at the top level do:
```
column_name_to_tuple_index = {}  # add type here
for tuple_index, column_name in enumerate(unexpected_index_column_names):
    if column_name not in index_names:
        raise ge_exceptions.MetricResolutionError(
            message=f"Error: The column {column_name} does not exist in the named indices. Please check your configuration",
             failed_metrics=["unexpected_index_list"],
        )
    column_name_to_tuple_index[column_name] = tuple_index
```
This does the validation early and  removes the call to `index` which is a hidden inner loop.

Then the loop here becomes:
```
for column_name in unexpected_index_column_names:
    primary_key_dict[column_name] = column_name_to_tuple_index[column_name]
```

Reply:
omg yes, thanks for the suggestion

Reply:
This didn't go as planned. The issue is that the necessary information is in two places:

`domain_records_df_index_names` contains the names of the named indices `[`pk1`, `pk_2`]` etc. 
`unexpected_indices` is the Tuple that contains the index values, but doesn't have the "actual" names of the columns.  So the `enumerate()` wasn't able to help us in this case

However, I moved the check that raises a `MetricResolutionError` outside of the big loop (`unexpected_index_column_names` is typically going to be pretty small), so we can error out earlier.